cpp/tests/test_metadata_payload_exchange.cpp

```diff
@@ -0,0 +1,300 @@
+            GlobalEnvironment->comm_, OpID{42}, statistics
+        );
+
+        GlobalEnvironment->barrier();
+    }
+
+    void TearDown() override {
+        GlobalEnvironment->barrier();
+    }
+
+    std::unique_ptr<MetadataPayloadExchange::Message> create_test_message(
+        Rank peer_rank, std::vector<std::uint8_t> metadata, std::size_t data_size = 0
+    ) {
+        std::unique_ptr<Buffer> data_buffer = nullptr;
+        if (data_size > 0) {
+            data_buffer =
+                br->allocate(stream, br->reserve_or_fail(data_size, MemoryType::DEVICE));
+            // Fill with test data
+            data_buffer->write_access(
+                [data_size](std::byte* ptr, rmm::cuda_stream_view stream) {
+                    std::vector<std::uint8_t> test_data(data_size);
+                    std::iota(test_data.begin(), test_data.end(), 0);
+                    RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
+                        ptr, test_data.data(), data_size, cudaMemcpyHostToDevice, stream
+                    ));
+                }
+            );
+            data_buffer->stream().synchronize();
+        }
+
+        return std::make_unique<MetadataPayloadExchange::Message>(
+            peer_rank, std::move(metadata), std::move(data_buffer)
+        );
+    }
+
+    std::unique_ptr<Buffer> allocate_receive_buffer(std::size_t size) {
+        return br->allocate(stream, br->reserve_or_fail(size, MemoryType::DEVICE));
+    }
+
+    void wait_for_communication_complete() {
+        auto allocate_fn = [this](std::size_t size) {
+            return allocate_receive_buffer(size);
+        };
+        for (int iter = 0; iter < 100 && !comm_interface->is_idle(); ++iter) {
+            comm_interface->receive_messages(allocate_fn);
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        }
+    }
+
+    Communicator* comm;
+    std::unique_ptr<rmm::mr::device_memory_resource> mr;
+    rmm::cuda_stream_view stream;
+    std::unique_ptr<BufferResource> br;
+    std::shared_ptr<Statistics> statistics;
+    std::unique_ptr<TagMetadataPayloadExchange> comm_interface;
+};
+
+TEST_F(MetadataPayloadExchangeTest, InitialState) {
+    // Communication interface should start in idle state
+    EXPECT_TRUE(comm_interface->is_idle());
+}
+
+TEST_F(MetadataPayloadExchangeTest, SendReceiveMetadataOnly) {
+    if (comm->nranks() < 2) {
+        GTEST_SKIP() << "Test requires at least 2 ranks";
+    }
+
+    Rank peer_rank = (comm->rank() + 1) % comm->nranks();
+    std::vector<std::uint8_t> test_metadata = {0x01, 0x02, 0x03, 0x04};
+
+    if (comm->rank() == 0) {
+        // Rank 0 sends metadata-only message
+        std::vector<std::unique_ptr<MetadataPayloadExchange::Message>> messages;
+        messages.push_back(create_test_message(peer_rank, test_metadata));
+
+        EXPECT_TRUE(comm_interface->is_idle());
+        comm_interface->send_messages(std::move(messages));
+        EXPECT_FALSE(comm_interface->is_idle());
```

Comment:
This expectation must be true because we haven't manually progressed the communication, is that right?